### DISCUSSION TITLE
What about `text suggestion` support?

### DISCUSSION THREAD

**Discussion by adriangalilea** [helix-editor/helix#10259] [General] [answered]
I fully understand the decision not to implement copilot...

But! Perhaps the interface of showing text suggestions can be on Helix so LSP's can use it.

<img width="480" alt="ghost suggestion text" src="https://github.com/helix-editor/helix/assets/90320947/cf9626e7-fa24-409a-9d8f-26e64c9bc799">

This is generic enough that maybe is better implemented as a primitive natively on Helix, and would improve the DX drastically of using any completion AI.

_Originally posted by @adriangalilea in https://github.com/helix-editor/helix/discussions/4037#discussioncomment-9012709_

**Comment by pascalkuthe** [comment]
lsp has completions which can support that, we are not adding lsp extensions to core

**Reply by adriangalilea** [reply]
Sorry, not sure if I understand what you are saying or if you misunderstood me.

Helix-gpt gives me this:
<img width="415" alt="image" src="https://github.com/helix-editor/helix/assets/90320947/2c52c91b-a1ec-483f-8abe-9606f4d91990">

Which is what I think you mean by completions.

What I'm saying is having baked in helix something that allows for inline suggestions/autocomplete text(dimmed) like I showed in the initial message.

This is something that can be leveraged by other any completion LSP, having a dropdown/window is fine for alternative options, but the inline text is the optimal DX.

Just  to clarify, I don't want copilot or any particular provider, what I want is a better interface for any completions flow.

**Reply by kirawi** [reply]
Yes but there isn't an extension for this behavior specifically. Maybe it could be implemented in a hacky way by the LSP by using inlay hints but I'm not sure.

**Reply by adriangalilea** [reply]
While copilot or other current llm completions providers may come and go, the interface of text suggestion will be ubiquitous.

I don't know if that level of smooth integration would be possible with plugins.

And even if it was, consider the amount of LLM completion providers there will exist, do you want to offload them the task of rebuilding the text suggestion inside the editor every time?

I think is worth considering having that baked into helix to allow that integration to whatever needs to use it more easily, perhaps create some sort of new standard.

> Maybe it could be implemented in a hacky way by the LSP by using inlay hints but I'm not sure.

I don't think LSP's are the way to go for this kind of feature, LSP's are for a specific purpose, that kind of text is for text suggestions.

I understand this is not easy and you guys have other priorities, I sympathise and agree with your negation to integrate copilot, but I want you to consider this.

**Reply by kirawi** [reply, marked as answer]
AFAIK the maintainers have no interest in implementing non-standard APIs like that. What you're proposing sounds pretty niche too. Stuff like that is intended to be implemented by plugin developers once the plugin system lands.

**Reply by gabydd** [reply]
Though we don't have some way to plug an external thing to render inline we do have the capabilities in tree to render inline for things like inlay hints and soft wrap so it should be possible to expose the APIs to plugins once we have that

**Reply by adriangalilea** [reply]
> AFAIK the maintainers have no interest in implementing non-standard APIs like that.
> What you're proposing sounds pretty niche too.

Could be, even tho it doesn't seem like it now, it'll be obvious in retrospect.

LLM's have been around a year, it's arguably the most rapid adoption on history, niche for now perhaps, standards are yet to be made. And if something is shaping up as a standard is that, you have it in gmail, you have it in copilot, and it feels very smooth.

> Stuff like that is intended to be implemented by plugin developers once the plugin system lands.

The majority of programmers will use AI, text suggestions is one of the nicest flows for it and relying on a plugin for something the majority will need may create friction on the adoption of new providers, thus my argument for having that baked in. Of course, that can be done via plugins, I'm just saying that it may be a better idea to have it in the core.

But it's alright, this is just my opinion, wanted to bring it up and see what other people thought.

Thanks for chiming in regardless

**Reply by kirawi** [reply]
As far as I've seen, AI has only taken off for webdev (it's unusable/unreliable for systems-level languages even for simple functions in my experience). I just don't see it being widely adopted by professional software engineers outside of that because of many reasons beyond just practical use, e.g. explainability which all LLMs lack at the moment (someone also has to verify that the AI wrote the code well, and at that point you may as well have written it yourself). One of my past professors who has her phd in ML and actively follows LLM papers emphasized those ethical liabilities/concerns which could lead to legal issues as major blockers for adoption. Maybe in a few decades something will come up with that ability, but perhaps AI can be left to the post-post-modern text editor at that point :P

(this is just my opinion)

**Reply by adriangalilea** [reply]
Have you used copilot in vscode?

Professional programmers use it already, the main blocker for some parts of the industry is precisely the subpar DX on alternative editors, copilot is not chatgpt.

I think you are confounding those, there's no need for explainability, copilot, in way too many instances feels as if it reads your mind, you just type out the start of a function, and in an instant you have that predictive text that more often than not is a perfect implementation of what you wanted to do.

I don't think there is any logical reason not to use it, it saves you way too much time on boilerplate and you can focus on tackling the real problems, and the main reason for how amazing it is is the text prediction interface that you can tab to complete or decide to ignore, is not intrusive.

**Reply by kirawi** [reply]
There is rarely boilerplate with systems-level programming as there is with webdev languages in my (limited) experience outside of Rust. I have tried copilot and yeah it doesn't do a great job. Sometimes it gets the types wrong or uses invalid syntax.

**Reply by pascalkuthe** [reply]
copilot gets way too much stuff wrong to be useable. It's more work for me to triple check whatever copiolt is writing (because its often very correct looking but not actually correct) than actually write stuff myself. Particularly because for most stuff I know what I want the code to look like and Copiolt never guesses to right version anyway.

Its similar with very junior devs: You spend more time checking and fixing stuff than it actually saves you. Just that junior devs learn but copiolt does not. 


When problems are complex, performance matters and bugs can be very subtle I most definitely don't want something like copiolt messing around in my code and producing subar results. It can be ok for webdev but I have 0 interested in using it (and I work with AI). And our general stance is that we don't add support for lsp extension or any kind of third party tooling into helix directly and we certainly won't make an exception for something non of us find particularly useful.

**Reply by kirawi** [reply]
It actually made me remember a time where I asked it write a function to swap certain characters in a string, and it wrote a function that just printed, `"All characters in the string have been swapped!"`

**Reply by adriangalilea** [reply]
> where I asked it write a function

This is why I'm not asking for a chat interface. If you are asking copilot it 99% does it wrong, for now.

@pascalkuthe 

Every single criticism you shared I agree and express myself, but just keep in mind, every tool has tradeoffs and you are judging **my proposal of an abstract interface** for the quality of the first generation tools that uses it.

I am not here to defend copilot, and debating it is missing the forest for the tree, but I'll say it's better than what you guys think, and disagreeing is ignoring the reality for most programmers now.

Anyway can't wait for plugins, to build that interface properly into helix :)

Thank you for what you are building, I love helix.

**Reply by ebetica** [reply]
Just wanted to add to this discussion, as a ML researcher and long time Kakoune user who's recently switched to VSCode (thankfully with the kakoune emulator plugin!), copilot has been a huge productivity boost. It's main uses do not lie in generating correct, complex, code, but rather abstracting out the boring bits of programming, where you have to match up function calls to the correct arguments or make some syntactically similar code snippets or function completions like:

```
a = do_something(a)
# copilot will now suggest
# b = do_something(b)
# c = do_something(c)

x = function_with_complex_arguments( #copilot will often suggest the correct arguments now
```

Essentially just snippets completion that actually works well. Another use case is in plotting and data visualization - since these syntactical transformations occur quite often + it's impossible to memorize the arguments in matplotlib.

I've actually been constantly looking for development on a kakoune like interface that implements LLM snippet suggestions and am quite disappointed to see this in Helix, as well as a lack of development in kakoune.

**Reply by kirawi** [reply]
Again, we would likely accept an LLM interface if there was an industry standard to implement. We are not interested in establishing our own standard because of the amount of work involved. That is why we are deferring this to the plugin system. In the mean time, the best option you have is to use either https://github.com/leona/helix-gpt or https://github.com/SilasMarvin/lsp-ai. If you're able to compile from source, you can also try the copilot fork at https://github.com/AlexanderDickie/helix

**Reply by adriangalilea** [reply]
> Again, we would likely accept an LLM interface if there was an industry standard to implement. We are not interested in establishing our own standard because of the amount of work involved. That is why we are deferring this to the plugin system.

Agreed for the most part, but I got something new to add...

I've heard in the [steel PR](https://github.com/helix-editor/helix/pull/8675) this comment by @merisbahti:

> I don't know if this is the place for this discussion, but I managed to install this version of helix and I wanna build a plugin.
> 
> I want to either make a "git blame" or "copilot suggestion" plugin - but both of them make use of some kind of ghost text and I've tried searching for it but cannot find a way to make ghost text.
> 
> I'm not sure if this is possible in this plugin-system today, or would it require some more capabilities provided from helix?

Which points again about the need I was trying to convey in this very issue. (Plugins are worthless without being able to display the `ghost text` thingy I was referring to)

**Then I've noticed this other comment by @jokeyrhyme :**

> I believe this is called "virtual text" within helix, so maybe patiently look out for any announcements or APIs relating to those words :)

@kirawi **couldn't Helix expose the `virtual text` control in a way that a plugin could use it?**, which is what I was proposing here from the beginning, I'm not asking for bespoke AI support. Again, all that a proper AI flow needs is a means to display the `ghost text` and a way to either accept(<kbd>tab</kbd>) or ignore it(keep writing). If `virtual text` exists already within helix then there's nothing to create, simply expose it.

P.S.:
>  In the mean time, the best option you have is to use either https://github.com/leona/helix-gpt or https://github.com/SilasMarvin/lsp-ai

Both are worse than not using anything, purely because of the interface, as explained before, do not waste your time, also, I wouldn't trust using a random repo with AI API keys blindly, so everyone beware.

**Reply by kirawi** [reply]
It's likely that it would be exposed to plugins, but the PR is an MVP for configuration so we can get it out in a reasonable state and work on it in subsequent PRs. It's already a lot to review.

**Reply by adriangalilea** [reply]
> It's likely that it would be exposed to plugins, but the PR is an MVP for configuration so we can get it out in a reasonable state and work on it in subsequent PRs. It's already a lot to review.

Agreed, focus on finishing it as is, but as you see, this was my petition from the very start, and is more reasonable than people thought, so this can be converted into an issue back again, or I can create a new one, so we can deal with it after the steel plugins PR is merged, so that the task is clear and the path to solve it too. I can, using the new knowledge about `virtual text` write it up properly and link to all mentions so people know where to go and what is needed to move this.

**Reply by pascalkuthe** [reply]
Issues are supposed to be actionable items that can be worked on. Hypothetical ideas for plugin APIs do not fall under that. You can *dicuss* them in a *discussion* but issues are supposed to be actionable. Until a plugin system is in place we don't need issues about adding features to it.

The virtual text API already exists internally (which should be rather obvious considering we have working inlay hints for a while now). So nothing is actionable regarding this in the codebase right now.

**Reply by adriangalilea** [reply]
> Issues are supposed to be actionable items that can be worked on. Hypothetical ideas for plugin APIs do not fall under that.

This is not hypothetical, I pointed you to [someone working on 2 different plugins(not only AI) that requires this already](https://github.com/helix-editor/helix/pull/8675#issuecomment-2235850288), as I suggested would happen months ago.

I don't think it's farfetched to have an issue that specifies X is blocked until Y is completed, as is meant to be fully actionable right after PR is merged. But as you wish, I can bump this idea after the merge...

> which should be rather obvious considering we have working inlay hints for a while now

So initially my proposal was outright rejected citing the need for non-standard APIs, and now it's rather obvious... pick one 😄 

Articulating this will make a very specific roadmap, hopefully toning down the "when AI copilot" comments

PR steel plugins merged -> virtual text API on steel plugins-> AI first class experience in any plugin(among other things)

Sounds like a plan, right?

**Reply by pascalkuthe** [reply]
> This is not hypothetical, 

That PR is in very early exploratory stages. The first step we will be merging is a scriptable config (for which there is an issue). The plugins implemented in that PR are proofs of concept and do not necessarily reflect the final API we will merge (which I discussed with the author).

We will work in multiple steps and virtual text API is realistically multiple steps away. It is too early to build a roadmap. 

> So initially my proposal was outright rejected citing the need for non-standard APIs, and now it's rather obvious... pick one 😄

Non-standard APIs was referring to the interaction with the LSP/ML tool not our internal text rendering engine API. 

We have more than enough issues and we are trying to reduce and centralize the number of open issues. Adding issues for APIs preconditioned on a plugin system that doesn't even exist yet is not helpful and would just get lost in the mass of 500 other issues. Hiding things that actually do need attention and are actionable right now. Even for features that are not preconditioned on anything else we usually close issues unless the feature has a detailed description of what that feature should do. A detailed description is hardly possible right now given that the prerequisites are not even in place.

**Reply by adriangalilea** [reply]
> We will work in multiple steps and virtual text API is realistically multiple steps away. It is too early to build a roadmap.

But the roadmap is already built, regardless of it's shape and how distant it is.

1. Plugins implementation, regardless if that PR or not.
2. `virtual text API in plugins` issue creation.
3. Implementing `virtual text API in plugins`.
4. ✅ Anyone can implement the perfect UX for AI.

> Non-standard APIs was referring to the interaction with the LSP/ML tool not our internal text rendering engine API.

Well, if you read me again, you'll realise that I was trying to convey that from the very beginning. In fact, with this new understanding, you'll see that the selected answer is missing the mark. (no offense)

> We have more than enough issues and we are trying to reduce and centralize the number of open issues.

Don't worry, let's do just that. I'll articulate a bit better in this discussion in a new comment, so everyone is clear.

@pascalkuthe @kirawi 

You don't have to convince me how hard it is to manage such a project, I appreciate and applaud the work you guys are doing.

I just wanted to offer my vision on how to solve this problem in a way that is compatible with your vision, that is, using plugins and keeping core minimal.

Regardless of how distant, being able to answer the question will be helpful, rather than a vague "with plugins" which will lead to incorrect expectations and a lack of understanding, as the very specific answer is, "with plugins, after virtual text API is implemented in them". Which I don't think it was obvious for anyone, since no one mentioned them in this conversation, and will not be an AI specific thing but an API that as discussed will be helpful for more types of plugins.

**Reply by karimknaebel** [reply]
Support for this [has been added to the LSP spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.18/specification/#textDocument_inlineCompletion) now to support the new official [Copilot language server](https://www.npmjs.com/package/@github/copilot-language-server).

**Reply by devmanuelli** [reply]
@adriangalilea @kabouzeid @pascalkuthe I'm implementing textDocument.inlineCompletion . By next monday I will have made a PR
EDIT done see here https://github.com/helix-editor/helix/pull/14876

**Comment by adriangalilea** [comment]
Closing the loop on this two years later, since the pieces finally exist:

First, credit where due: @devmanuelli already implemented `textDocument/inlineCompletion` in #14876, and pinged me here last November. Sorry for the slow reply, I missed it. That is the right layer for this, and it looks close: two approvals, and the `LineAnnotation` / `Decoration` feedback addressed. From the outside it seems to mostly need a rebase.

- LSP 3.18 standardized this exact interface as [`textDocument/inlineCompletion`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.18/specification/#textDocument_inlineCompletion), so "the interface of showing text suggestions" is no longer an LSP extension, it's the spec. See also #14059 asking for it.
- As an experiment I built the primitive I was describing here, at plugin level on the Steel fork: a ghost-text slot (show / clear / accept, plus a pending state so a streaming suggestion can't be accepted half-written) rendered through the existing annotation machinery, with the suggestion source living entirely in a plugin. It's a rough prototype on a personal branch (I'm daily-driving it, not proposing it), but it does show the split is practical: https://github.com/adriangalilea/helix/blob/steel%2Bfilesentry/GHOST-TEXT.md

<img width="3274" height="1975" alt="image" src="https://github.com/user-attachments/assets/e926d44c-427d-4462-b5c5-f109421877f2" />
This is how it looks

One datapoint for #14876: I arrived independently at the same `LineAnnotation` + `Decoration` shape archseer asked for there, so that approach does hold up in practice.

Not re-litigating core inclusion, just leaving the trail for whoever lands inlineCompletion support: the rendering primitive and the protocol are cleanly separable, and separating them is what made this buildable outside core at all.
